File: eel-core/src/main/kotlin/io/eels/component/json/JsonSource.kt
package io.eels.component.json

import io.eels.Row
import io.eels.Source
import io.eels.component.Part
import io.eels.component.Using
import io.eels.schema.Field
import io.eels.schema.Schema
import io.eels.util.map
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.fs.FSDataInputStream
import org.apache.hadoop.fs.FileSystem
import org.apache.hadoop.fs.Path
import org.codehaus.jackson.JsonNode
import org.codehaus.jackson.map.ObjectMapper
import org.codehaus.jackson.map.ObjectReader
import rx.Observable

data class JsonSource(val path: Path) : Source, Using {

  private val reader: ObjectReader = ObjectMapper().reader(JsonNode::class.java)

  private fun createInputStream(path: Path): FSDataInputStream {
    val fs = FileSystem.get(Configuration())
    return fs.open(path)
  }

  override fun schema(): Schema = using(createInputStream(path)) {
    val roots = reader.readValues<JsonNode>(it)
    val node = roots.asSequence().first()
    val fields = node.fieldNames.asSequence().map { Field(it) }.toList()
    Schema(fields)
  }

  override fun parts(): List<Part> = listOf(JsonPart(path))

  inner class JsonPart(val path: Path) : Part {

    val schema = schema()

    fun nodeToRow(node: JsonNode): Row {
      val values = node.elements.map { it.textValue }.asSequence().toList()
      return Row(schema, values)
    }

    override fun data(): Observable<Row> = Observable.create<Row> { sub ->
      val input = createInputStream(path)
      try {
        sub.onStart()
        reader.readValues<JsonNode>(input).asSequence().forEach {
          val row = nodeToRow(it)
          sub.onNext(row)
        }
      } catch (t: Throwable) {
        sub.onError(t)
      }
      sub.onCompleted()
    }
  }
}

File: eel-core/src/main/scala/io/eels/component/json/JsonSource.scala
package io.eels.component.json

import java.io.InputStream
import java.nio.file.Files
import java.util.concurrent.atomic.AtomicBoolean

import com.sksamuel.exts.io.Using
import io.eels._
import io.eels.datastream.{DataStream, Publisher, Subscriber, Subscription}
import io.eels.schema.{ArrayType, DataType, Field, StructType}
import org.apache.hadoop.fs.{FileSystem, Path}
import org.codehaus.jackson.JsonNode
import org.codehaus.jackson.map.{ObjectMapper, ObjectReader}

import scala.collection.JavaConverters._

// the stream must be repeatable. That is, the input function may be called multiple times.
case class JsonSource(inputFn: () => InputStream,
                      inferrer: SchemaInferrer = StringInferrer) extends Source with Using {

  def withSchemaInferrer(inferrer: SchemaInferrer): JsonSource = copy(inferrer = inferrer)

  private val reader: ObjectReader = new ObjectMapper().reader(classOf[JsonNode])

  private def field(name: String, node: JsonNode): Field = {
    node match {
      case obj if obj.isObject => Field(name, struct(obj))
      case arr if arr.isArray => Field(name, ArrayType(datatype(name, arr.iterator().next)))
      case _ => inferrer.infer(name)
    }
  }

  private def datatype(name: String, node: JsonNode): DataType = {
    node match {
      case obj if obj.isObject => struct(obj)
      case arr if arr.isArray => ArrayType(datatype(name, arr.iterator().next))
      case _ => inferrer.infer(name).dataType
    }
  }

  private def struct(obj: JsonNode): StructType = StructType(obj.getFields.asScala.map { entry =>
    field(entry.getKey, entry.getValue)
  }.toVector)

  // we take the schema from the first entry only if there are several
  lazy val schema: StructType = using(inputFn()) { in =>
    val roots = reader.readValues[JsonNode](in)
    assert(roots.hasNext, "Cannot read schema, no data in file")
    struct(roots.next)
  }

  override def parts(): Seq[Publisher[Seq[Row]]] = List(new JsonPublisher(inputFn))

  class JsonPublisher(inputFn: () => InputStream) extends Publisher[Seq[Row]] {

    private def nodeToValue(node: JsonNode): Any = {
      if (node.isArray) {
        node.getElements.asScala.map(nodeToValue).toVector
      } else if (node.isObject) {
        node.getFields.asScala.map { entry =>
          entry.getKey -> nodeToValue(entry.getValue)
        }.toMap
      } else if (node.isBinary) {
        node.getBinaryValue
      } else if (node.isBigInteger) {
        node.getBigIntegerValue
      } else if (node.isBoolean) {
        node.getBooleanValue
      } else if (node.isTextual) {
        node.getTextValue
      } else if (node.isLong) {
        node.getLongValue
      } else if (node.isInt) {
        node.getIntValue
      } else if (node.isDouble) {
        node.getDoubleValue
      }
    }

    private def nodeToRow(node: JsonNode): Row = {
      val values = node.getElements.asScala.map(nodeToValue).toArray
      Row(schema, values)
    }

    override def subscribe(subscriber: Subscriber[Seq[Row]]): Unit = {
      try {
        using(inputFn()) { input =>
          val running = new AtomicBoolean(true)
          subscriber.subscribed(Subscription.fromRunning(running))
          reader.readValues[JsonNode](input).asScala
            .takeWhile(_ => running.get)
            .map(nodeToRow)
            .grouped(DataStream.DefaultBatchSize)
            .foreach(subscriber.next)
          subscriber.completed()
        }
      } catch {
        case t: Throwable => subscriber.error(t)
      }
    }
  }
}

object JsonSource {

  def apply(path: Path)(implicit fs: FileSystem): JsonSource = {
    require(fs.exists(path), s"$path must exist")
    JsonSource(() => fs.open(path))
  }

  def apply(path: java.nio.file.Path): JsonSource = {
    require(path.toFile.exists, s"$path must exist")
    JsonSource(() => Files.newInputStream(path))
  }
}
